ports/esp32-esp8266: Add support for set/get the wifi power saving mode.

glenn20 opened the pull request:
This PR adds support for setting and getting the wifi power saving mode.

Adds the `pm` option to `WLAN.config()` to support setting/getting the wifi power saving mode.

Also adds the `PM_NONE`, `PM_MIN_MODEM` and `PM_MAX_MODEM` constants to the `WLAN` class.

Example:
```
import network
sta = network.WLAN(0); sta.active(True)
sta.connect(ssid, password)
sta.config(pm=sta.PM_NONE)
```

This is especially useful for controlling power consumption on battery powered devices and is also required by PR #6515 (ESPNow support).

peterhinch commented:
This sounds similar to [functionality in the esp module](http://docs.micropython.org/en/latest/library/esp.html#esp.sleep_type). Please could you explain the relationship between the two.

glenn20 commented:
> This sounds similar to [functionality in the esp module](http://docs.micropython.org/en/latest/library/esp.html#esp.sleep_type). Please could you explain the relationship between the two.

@peterhinch Good pickup - thanks Peter. I was not aware this functionality was available in the esp module (ESP8266 only).

I confess that I find it odd that this one wifi option is set in the esp module and everything else is set in WLAN.config(). It is hardly surprising that I and other users have overlooked this till now. Nonetheless, precedent is an important factor.

I guess I could add support for the ESP32 to ports/esp32/modesp.c. However, I really do think users are going to continue to overlook this feature if it is not in WLAN.config().

Thoughts?

dpgeorge commented:
I think settings should be consolidated to `WLAN.config()`.  Older functions like `phy_mode()` and `sleep_type()` were added before `.config()` was introduced.

But we also need to standardise on naming for setting the power saving mode.  In the CYW43 driver it uses `.config(pm=...)` (for power-management).

glenn20 commented:
@dpgeorge Ok - I can add a commit renaming the option to "pm" to be consistent with the cyw43 driver. 

I am assuming it is too difficult to attempt to standardise the values (and constant names) for those arguments as the semantics are likely to vary considerably across devices (even between the esp32 and esp8266).

dpgeorge commented:
I think the constants need to go inside the WLAN class itself.  Because if there are multiple NICs then they could all have different options/values for the power saving configuration.

I would suggest naming the constants `PM_xxx`: the "PM" matches the name of the config argument, and it doesn't need to be prefixed with `WIFI_` if it lives in the WLAN class.

glenn20 commented:
> I think the constants need to go inside the WLAN class itself. Because if there are multiple NICs then they could all have different options/values for the power saving configuration.
> 
> I would suggest naming the constants `PM_xxx`: the "PM" matches the name of the config argument, and it doesn't need to be prefixed with `WIFI_` if it lives in the WLAN class.

Ok. Shall do.

@dpgeorge : This may start to create a consistency issue with respect to many existing constants in the network module which would by the same logic belong in the WLAN class. I'm very happy that we can start with just this set of constants (which I agree makes sense in the light of wider network adaptor support in micropython) - just flagging that it will seem a bit arbitrary to users which constants are in network and which in WLAN().

glenn20 commented:
I have rebased against master and pushed a commit to implement (as requested):

- Move WIFI_PS_ constants from network module to WLAN() class.
- Rename WIFI_PS_* constants to PM_*

for esp32 and esp8266.

glenn20 committed:
esp32,esp8266: Add support to set/get power saving mode of WLAN.

For esp32 and esp8266 this commit adds:
- a 'pm' option to WLAN.config() to set/get the wifi power saving mode; and
- PM_NONE, PM_PERFORMANCE and PM_POWERSAVE constants to the WLAN class.

This API should be general enough to use with all WLAN drivers.

Documentation is also added.

glenn20 commented:
Whoops - I forgot to rename the PS_ constants on esp32. Fixed now.

AmirHmZz commented:
Any updates on this one?

glenn20 commented:
> Any updates on this one?

No updates. Just waiting for further review.

dakoner commented:
FWIW I patched this and it worked great; my ping times go from 100ms to 2-3ms.

Can you update the first comment to show the current configuration/code example, it should be:

import network
sta = network.WLAN(0)
sta.active(True)
sta.connect(ssid, password)
sta.config(pm=sta.PM_NONE)

glenn20 commented:
@dakoner - Glad it worked for you and thanks for the tips to fix the docs. Done.

jjsuwa-sys3175 commented:
This is just what I want and searched, because WebREPL is unbearably slow in the current situation.
I think the negative side of power savings can get in the way when trying out various things.

I wonder why it's still not committed...

glenn20 commented:
Rebased against master to resolve conflicts.

glenn20 commented:
> Does this PR allow wifi/bt modems to be completely switched off as described as "modem sleep" here?.

This looks like the PS_MIN_MODEM mode described [here](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-guides/wifi.html#esp32-wi-fi-power-saving-mode).

That wifi ps mode is automatically enabled when the esp32 connects to a wifi access point. You can use this PR to doable this power saving mode or select the high power savings of PS_MAX_MODEM.

dpgeorge commented:
We need to also consider WiFi modules based on the cyw43.  The existing `extmod/network_cyw43.c` driver actually includes an (undocumented) `wlan.config(pm=...)` option.  It has 3 predefined power-management settings called `DEFAULT_PM`, `AGGRESSIVE_PM` and `PERFORMANCE_PM` (which aren't yet exposed to Python).  But you can actually pass an integer value to the `pm` option and you get quite a lot of control over how active the WiFi is:
- no power saving (never turn off) vs full power saving (turn off whenever possible) vs balanced (turn off after some configurable time T)
- the association interval sent to the AP (ie how long the AP should wait between pings of the STA)
- how often the STA should wake up to listen for frames from the AP

@glenn20 do you know if you have such control over the WiFi power management in the esp32?  Or, what do the PS_MIN_MODEM/MAX_MODEM settings actually correspond to in terms of the above parameters?

glenn20 commented:
> We need to also consider WiFi modules based on the cyw43. The existing `extmod/network_cyw43.c` driver actually includes an (undocumented) `wlan.config(pm=...)` option. It has 3 predefined power-management settings called `DEFAULT_PM`, `AGGRESSIVE_PM` and `PERFORMANCE_PM` (which aren't yet exposed to Python). But you can actually pass an integer value to the `pm` option and you get quite a lot of control over how active the WiFi is:
> 
> * no power saving (never turn off) vs full power saving (turn off whenever possible) vs balanced (turn off after some configurable time T)
> * the association interval sent to the AP (ie how long the AP should wait between pings of the STA)
> * how often the STA should wake up to listen for frames from the AP
> 
> @glenn20 do you know if you have such control over the WiFi power management in the esp32? Or, what do the PS_MIN_MODEM/MAX_MODEM settings actually correspond to in terms of the above parameters?

The ESP32 implements only the wifi modem-sleep mode as described at [ESP32 Wifi Power Saving Mode](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-guides/wifi.html#esp32-wi-fi-power-saving-mode):

- `WIFI_PS_MIN_MODEM`: Wakes every DTIM to receive beacon. The DTIM interval is set by the AP. No data is lost.
- `WIFI_PS_MAX_MODEM`: Wakes up every "listen interval". The listen interval can be set in the wifi config. It can be set longer than the AP's DTIM interval, but data may be lost. (NOTE. setting this listen interval is not exposed by micropython).
- `WIFI_PS_NONE`: Does not shut down radio while connected to AP.

which "conceptually" map to the modes in the cyw43 driver. (Note that the ESP32 automatically enters `WIFI_PS_MIN_MODEM` when it connects to an AP - hence the initial motivation for this PR).

However, looking at the docs for `cyw43_pm_value()` (used to encode that integer argument) it must be using some other 802.11 PM extensions beyond modem-sleep, as the docs imply the client tells the AP what listen interval to use for this client.

The signature for the "pm" config option in this PR is compatible with the cyw43 driver, BUT the meaning of the values of the arg are different. Are you thinking to find a common base naming convention for a few constants that map to the port/device specific settings (eg. `PM_NONE`, `PM_MIN`, `PM_MAX`)?

Also, we **could** consider encoding the "listen interval" into the "pm" argument like `cyw43_pm_value()` does, to give more control, but I don't think the demands is high for that at this stage.

glenn20 commented:
Aside: The cyw43 docs suggest that DEFAULT_PM saves more power than AGGRESSIVE_PM, which is not what I "expected" from the naming convention.

dpgeorge commented:
> Are you thinking to find a common base naming convention for a few constants that map to the port/device specific settings

At the very least, yes.

> Also, we **could** consider encoding the "listen interval" into the "pm" argument like `cyw43_pm_value()` does, to give more control

I would like to do this.  Eg there could be a few options like `pm_mode`, `pm_interval`, `pm_dtim`.  Or a single `pm` option that takes an opaque integer which:
- there are default values provided as constants, eg `WLAN.PM_PERFORMANCE`
- there's a function used to construct a value, eg `WLAN.pm_value(mode, dtim)`

In most cases users would just do `wlan.config(pm=WLAN.PM_PERFORMANCE)`, but they could also do `wlan.config(pm=WLAN.pm_value(...))`.

dpgeorge commented:
> The cyw43 docs suggest that DEFAULT_PM saves more power than AGGRESSIVE_PM, which is not what I "expected" from the naming convention

Yes, the naming of the default constant is poor.  We can change that.

glenn20 commented:
> I would like to do this. Eg there could be a few options like `pm_mode`, `pm_interval`, `pm_dtim`. Or a single `pm` option that takes an opaque integer which:
> 
> * there are default values provided as constants, eg `WLAN.PM_PERFORMANCE`
> * there's a function used to construct a value, eg `WLAN.pm_value(mode, dtim)`
> 
> In most cases users would just do `wlan.config(pm=WLAN.PM_PERFORMANCE)`, but they could also do `wlan.config(pm=WLAN.pm_value(...))`.

I like this. You get a simple, portable method for setting common wifi PM settings and an extension api for tuning (which may be supported by only some drivers and may have different signatures and semantics on different drivers). 

(Pauses to think....) However, is this the cleanest method? If a `WLAN.pm_value()` method is being introduced, wouldn't it be simpler to just have a `WLAN.set_pm()` extension method to set the PM value, rather than encoding settings into an `int` and then passing that to `config()`. (ie. `config(pm=..)` for portable api and `set_pm(...)` for extension api.

dpgeorge commented:
> * Or `config(pm=n)` for portable and `config(pm=(tuple...))` for tuning extension.

Yes, I like that idea.

glenn20 commented:
> > * Or `config(pm=n)` for portable and `config(pm=(tuple...))` for tuning extension.
> 
> Yes, I like that idea.

Ok. Just need a naming convention for portable constants. I suggest a "none", "default" and 2-3 power level settings, eg: PM_NONE and PM_DEFAULT as well as something like:
- PM_PERFORMANCE, PM_POWERSAVE or
- PM_MIN_POWER, PM_MED_POWER, PM_MAX_POWER

PM_NONE - disable power saving
PM_DEFAULT - restores the default power saving setting of the chip/driver.

Let me know what you like. I'll update this PR with performance/powersave/default/none but can change at any time. I'll also investigate supporting the "listen interval". If I get the time this weekend, I'll test with the PPK2.

glenn20 commented:
Now supports:
```py
WLAN.config(pm=WLAN.PM_NONE)  # WIFI_PS_NONE
WLAN.config(pm=WLAN.PM_DEFAULT)  # WIFI_PS_MIN_MODEM
WLAN.config(pm=WLAN.PM_PERFORMANCE)  # WIFI_PS_MIN_MODEM
WLAN.config(pm=WLAN.PM_POWERSAVE)  # WIFI_PS_MAX_MODEM
WLAN.config(pm=(WLAN.PM_POWERSAVE, listen_interval))  # WIFI_PS_MAX_MODEM and set listen_address parameter in sta config.
```

Happy to change-up constant names or anything else. It might be helpful to have standardised named for 3 PM_ levels (this scheme only has 2).

dpgeorge commented on docs/library/network.WLAN.rst:
```diff
@@ -133,4 +133,24 @@ Methods
+
+.. data:: WLAN.PM_PERFORMANCE
+        WLAN.PM_POWERSAVE
+        WLAN.PM_DEFAULT
```
I suggest removing `PM_DEFAULT`.  I don't think it's useful and adds unnecessarily to the code size.

glenn20 replied:
Done

dpgeorge commented on docs/library/network.WLAN.rst:
```diff
@@ -133,4 +133,24 @@ Methods
+
+        * ``PM_PERFORMANCE``: high performance wifi with higher power consumption
+        * ``PM_POWERSAVE``: reduce wifi power consumption with reduced
+          performance
```
For performance it should be clear that it's still doing PM.  I suggest something like "enable power management but retain performance of the WiFi by only sleeping the RF components after a period of inactivity", or "balanced power management and WiFi performance"

For powersave, "enable maximum power saving capabilities at the expense of reduced WiFi performance".

glenn20 replied:
Good point. I hope I've respected your suggestion: 
- "enable WiFi power management to balance power savings and WiFi performance" and 
- "enable WiFi power management with additional power savings and reduced WiFi performance"
(leaving out any specifics about how the power saving is achieved - an inactivity timer is one way)

dpgeorge commented on docs/library/network.WLAN.rst:
```diff
@@ -133,4 +133,24 @@ Methods
+        * ``PM_DEFAULT``: default power settings used by the wifi chip.
+
+    Some wifi chip drivers may also allow setting the *pm* config option to a
+    tuple of driver-specific parameters (for fine tuning power management).
```
I suggest to remove the tuple capability for now.  We can add it later if needed.

glenn20 replied:
Done

dpgeorge commented on ports/esp32/network_wlan.c:
```diff
@@ -508,6 +508,24 @@ STATIC mp_obj_t network_wlan_config(size_t n_args, const mp_obj_t *args, mp_map_
+                    case MP_QSTR_pm: {
+                        mp_obj_t arg = kwargs->table[i].value;
+                        mp_int_t pm_mode;
+                        if (!mp_obj_get_int_maybe(arg, &pm_mode)) {
```
I suggest to remove this capability for now, and we can later work out how to make it compatible across WiFi drivers.

dpgeorge commented on ports/esp32/network_wlan.c:
```diff
@@ -627,6 +637,12 @@ STATIC const mp_rom_map_elem_t wlan_if_locals_dict_table[] = {
+
+    // Constants
+    { MP_ROM_QSTR(MP_QSTR_PM_NONE), MP_ROM_INT(WIFI_PS_NONE) },
+    { MP_ROM_QSTR(MP_QSTR_PM_DEFAULT), MP_ROM_INT(WIFI_PS_MIN_MODEM) },
```
This PM_DEFAULT constant needs to be removed.

dpgeorge commented on ports/esp8266/network_wlan.c:
```diff
@@ -511,6 +519,12 @@ STATIC const mp_rom_map_elem_t wlan_if_locals_dict_table[] = {
+
+    // Constants
+    { MP_ROM_QSTR(MP_QSTR_PM_NONE), MP_ROM_INT(NONE_SLEEP_T) },
+    { MP_ROM_QSTR(MP_QSTR_PM_DEFAULT), MP_ROM_INT(MODEM_SLEEP_T) },
```
As above, please remove.

glenn20 replied:
Thanks for picking this up - I think I stuffed up the commit. Fixed now.

dpgeorge commented:
Thanks, that looks better.

I squashed and rebased this PR.

dpgeorge commented:
Merged!

A concise PR is always best.

cyberedgelord commented:
Thank you for fixing this!

This one line:
`network.WLAN(network.STA_IF).config(pm=network.WLAN(network.STA_IF).PM_NONE)`

replaces my entire "workaround" on ESP8266 for ESP-NOW:
```
async def _event_keep_alive(self):
    while True:
        await self._on_sta_connected.wait()
        network.WLAN(network.STA_IF).connect()
        await uasyncio.sleep_ms(self._DTIM)

```

I noticed, that while you ping ESP8266, it can receive ESP-NOW messages, but once you stop, power saving mode kicks in and packets be gone.

dpgeorge commented:
You can simplify that one line to:
```
network.WLAN(network.STA_IF).config(pm=network.WLAN.PM_NONE)
```

glenn20 commented:
> You can simplify that one line to:
> ```
> network.WLAN(network.STA_IF).config(pm=network.WLAN.PM_NONE)
> ```

That does not work anymore - since network.WLAN was replaced by a function to return the WLAN instances rather than being a class in the network module.

I usually use:
```py
sta = network.WLAN(network.STA_IF)
sta.config(pm=sta.PM_NONE)
```

dpgeorge commented:
> That does not work anymore - since network.WLAN was replaced by a function to return the WLAN instances rather than being a class in the network module.

Oh, you're right!  And that's been there since the beginning, originating on esp8266.

That should definitely be fixed, to match other ports and network classes.

dpgeorge commented:
> That should definitely be fixed, to match other ports and network classes.

See #11465.
